provider/ec2/instancetypes.go

```diff
@@ -173,26 +174,39 @@ func archName(in string) string {
 	return in
 }
 
-func (e *environ) supportedInstanceTypes(ec2Session ec2iface.EC2API, ctx context.ProviderCallContext) (result []instances.InstanceType, err error) {
+func (e *environ) supportedInstanceTypes(ec2Session ec2iface.EC2API, ctx context.ProviderCallContext) ([]instances.InstanceType, error) {
 	e.instTypesMutex.Lock()
 	defer e.instTypesMutex.Unlock()
 
-	defer func() {
-		if err == nil {
-			if len(e.instTypes) == 0 {
-				logger.Tracef("got instance types:\n%#v", result)
-			}
-			e.instTypes = result
-		}
-	}()
-
 	// Use a cached copy if populated as it's mildly
 	// expensive to fetch each time.
 	// TODO(wallyworld) - consider using a cache with expiry
-	if len(e.instTypes) > 0 {
-		return e.instTypes, nil
+	if len(e.instTypes) == 0 {
+		instTypes, err := e.collectSupportedInstanceTypes(ec2Session, ctx)
+		if err != nil {
+			return nil, err
+		}
+		e.instTypes = instTypes
+
+		var filtList []instances.InstanceType
+		for _, it := range e.instTypes {
+			if !it.Deprecated && it.Cost < math.MaxUint64 {
```

Comment:
I think from what @jameinel was saying; we do want to include the `math.MaxUint64` because we're using spot instance types as a corollary for on-demand prices (which is totally wrong, but we have what we have), then you do want to include the instance types which have no cost.

That way you can say I want an instance type with 128GB, you'll end up with an instance type, that might not have a cost, but you do at least end up with an instance of said constraints.

The real fix here is to find better pricing API calls, yet I believe that's expensive in terms of the number of API calls.

Reply:
@SimonRichardson ah crap! that was a debug log from my tests and should not be in there :-(
Sorry, let me amend that.